server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -485,13 +486,29 @@ public void deleteShardDirectoryUnderLock(ShardLock lock, IndexSettings indexSet
         assert assertPathsDoNotExist(paths);
     }
 
-    private static boolean assertPathsDoNotExist(final Path[] paths) {
+    private static boolean assertPathsDoNotExist(final Path[] paths) throws IOException {
         Set<Path> existingPaths = new HashSet<>();
         for (Path path : paths) {
             if (FileSystemUtils.exists(path)) {
                 existingPaths.add(path);
             }
         }
+        // Relaxed assertion for the special case where only the empty state directory exists after deleting
+        // the shard directory because it was created again as a result of a metadata read action concurrently.
+        if (existingPaths.size() == 1) {
```

Comment:
I wonder if it's easier to move the "check if there's only a left-over `MetaDataStateFormat.STATE_DIR_NAME`" logic into the `if (FileSystemUtils.exists(path)) {` condition above. Then you won't have to remove stuff again from `existingPaths`.

Reply:
yea that would save a few lines :) sec adjusting